File: src/Scripting/Lua/patterns/visitor.lua
local nodes = { { kind="number", value=2 }, { kind="text", value="hi" } }
local visitor = {
  number = function(node) return node.value * 2 end,
  text = function(node) return string.upper(node.value) end
}
assert(visitor[nodes[1].kind](nodes[1]) == 4)
assert(visitor[nodes[2].kind](nodes[2]) == "HI")
return true

File: src/Scripting/PHP/patterns/visitor.php
<?php declare(strict_types=1);
$check = static fn(bool $ok) => $ok ?: throw new RuntimeException('visitor');
$nodes = [['type'=>'text','value'=>'Hi'], ['type'=>'number','value'=>3]];
$visitor = [
    'text' => static fn(array $n): string => $n['value'],
    'number' => static fn(array $n): string => (string)($n['value'] * 2),
];
$out = array_map(static fn($n) => $visitor[$n['type']]($n), $nodes);
$check($out === ['Hi','6']);
